controlplane/pkg/nsmd/nsmd_crossconnect_client.go

```diff
@@ -428,39 +420,45 @@ func (client *NsmMonitorCrossConnectClient) handleXconEvent(event monitor.Event,
 	return nil
 }
 
-func (client *NsmMonitorCrossConnectClient) remotePeerConnectionMonitor(ctx context.Context, remotePeer *registry.NetworkServiceManager) {
-	span := spanhelper.FromContext(ctx, fmt.Sprintf("remotePeerMonitor-%v", remotePeer.Name))
+func (client *NsmMonitorCrossConnectClient) remotePeerConnectionMonitor(ctx context.Context, remotePeer RemotePeerDescriptor) {
+	client.remotePeerLock.Lock()
+	remoteNsm := remotePeer.RemoteNsm()
+	client.remotePeerLock.Unlock()
+	span := spanhelper.FromContext(ctx, fmt.Sprintf("remotePeerMonitor-%v", remoteNsm.Name))
 	defer span.Finish()
 	grpcConnectionSupplier := func() (*grpc.ClientConn, error) {
-		span.Logger().Infof(peerLogWithParamFormat, remotePeer.Name, "Connecting to", remotePeer.Url)
-		return tools.DialContextTCP(span.Context(), remotePeer.GetUrl())
+		span.Logger().Infof(peerLogWithParamFormat, remoteNsm.Name, "Connecting to", remoteNsm.Url)
+		return tools.DialContextTCP(span.Context(), remoteNsm.GetUrl())
 	}
 	monitorClientSupplier := func(conn *grpc.ClientConn) (monitor.Client, error) {
 		return connectionMonitor.NewMonitorClient(conn, &connection.MonitorScopeSelector{
 			NetworkServiceManagers: []string{
 				client.xconManager.GetNsmName(), // src
-				remotePeer.Name,                 // dst
+				remoteNsm.Name,                  // dst
 			},
 		})
 	}
 
 	err := client.monitor(
 		ctx,
-		peerLogFormat, peerLogWithParamFormat, remotePeer.Name,
+		peerLogFormat, peerLogWithParamFormat, remoteNsm.Name,
 		grpcConnectionSupplier, monitorClientSupplier,
-		client.handleRemoteConnection, nil, map[string]string{peerName: remotePeer.Name})
+		client.handleRemoteConnection, nil, map[string]string{peerName: remoteNsm.Name})
 
 	if err != nil {
 		span.LogError(err)
 		// Same as DST down case, we need to wait for same NSE and updated NSMD.
-		connections := client.xconManager.GetClientConnectionByRemote(remotePeer)
+		connections := client.xconManager.GetClientConnectionByRemote(remoteNsm)
 		for _, cc := range connections {
 			ccSpan := common.SpanHelperFromConnection(ctx, cc, "remotePeerConnectionMonitor.update")
 			defer ccSpan.Finish()
 			ccSpan.LogObject("clientConnection", cc)
 			client.xconManager.DestinationDown(ccSpan.Context(), cc, true)
 		}
 	}
+	client.remotePeerLock.Lock()
```

Comment:
`remotePeerConnectionMonitor` in charge of only one remote peer, why it locks the whole map of peers?

Reply:
Peer canceling should be synchronized with Adding/Removing client connection

Reply:
That's why makes sense to add locks inside each RemotePeerDescriptor

Reply:
thanks, fixed.